### ISSUE TITLE
Investigate on how to support maxPageSize in RLC and Modular

### ISSUE THREAD

**Issue by MaryGao** [Azure/autorest.typescript#2802]
As offline discussed we plan to remove the core-paging dependency and maxPageSize support in RLC and modular. But we still have the requirement to support maxPageSize for some services, in this issue we need to investigate on how to support maxPageSize in RLC and Modular.

Relevant issues:
- https://github.com/Azure/autorest.typescript/issues/2780
- https://github.com/Azure/autorest.typescript/issues/2758
- https://github.com/Azure/autorest.typescript/issues/2696

**Comment by MaryGao** [comment]
We received this requirement for maxPageSize here: https://github.com/Azure/azure-sdk-for-js/pull/31848#discussion_r1854866602. So I add the priority to p0.

**Comment by MaryGao** [comment]
**Option 1: Accept this breakings**

```ts
// before
for await (const page of client
      .listPropertiesOfKeyVersions(keyName)
      .byPage({ maxPageSize: 1 })) { ... }

// now 
for await (const page of client
      .listPropertiesOfKeyVersions(keyName, { maxresults: 1 }) // some services may be top or n or pagesize etc
      .byPage()) { ... }
```

**Option 2: Provide a option to enable the maxPageSize in PageSetting + Customization**

Add a new option `IncludeMaxPageSize` for paging operations and then directly generating 

```ts
export interface PageSettings {
  /**
   * A reference to a specific page to start iterating from.
   */
  continuationToken?: string;
  /**
   * The size of the page during paged iteration
   */
  maxPageSize?: number;
}
```

The generated code would throw exceptions when handling `maxPageSize` so service team needs to override their mappings between maxPageSize option and real query parameter.

**Comment by MaryGao** [comment]
Conclusion for [above discussion](https://github.com/Azure/autorest.typescript/issues/2802#issuecomment-2497449921).
- KeyVault may not accept the breakings and we could work with them to re-write the paginate helper to support maxPageSize;
- Codegen could keep it as it is and collect more evidence to decide if we need an option to support this for backward compatibility reasons;
- maxPageSize will not take effective if we consume a full url of continuationToken which is required with our current guideline so that would not make much sense. But if continuationToken is not a full uri, and it could work properly with maxPageSize in byPage option, there may be useful to provide this option.

**Comment by maorleger** [comment]
Thanks for digging into this. I think for KeyVault we will need to continue to support the maxPageSize operation. I am also curious why it was removed in the first place given that it is part of the [REST API guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#maxpagesize) for services - won't our codegen need to continue supporting this for services that may provide this?

But, I understand I don't have all the context here so I am happy to focus on KeyVault only - for KeyVault we will need to continue supporting maxPageSize and ideally that would be codegen'd via option 1 (a backwards compatibility flag if I understood correctly).

**Comment by timovv** [comment]
The issue is that there is a difference between the max page size being specified at a per-page level:

(1):

```ts
client.listPropertiesOfSecrets().byPage({ maxPageSize });
```

and the max page size being specified at the first call to the operation:

(2):
```ts
client.listPropertiesOfSecrets({ maxPageSize }).byPage();
```

My understanding is that the API guidelines are advocating for is something closer to (2), and Key Vault actually implements it this way in its specification, albeit with a different name (the `maxresults` parameter). The generator supports (2) and generates `maxresults` at the operation level correctly. But the convenience library exposes (1) to customers. My understanding is we are hesitant to provide support for (1) at the codegen level since this isn't a recommended pattern for new libraries. 

For Key Vault I think we can paper over this with a helper to lazily call the generated operation so that the `maxPageSize` passed in through `byPage` is promoted and used to populate the `maxresults` property on the operation. We could probably do this if we extend the `mapPagedAsyncIterable` util we are already using to do that as well. Something like this (although I haven't tested it):

```ts
import type { PagedAsyncIterableIterator as GeneratedPagedAsyncIterableIterator } from "./generated/index.js";
import type { PagedAsyncIterableIterator as CorePagedAsyncIterableIterator } from "@azure/core-paging";

export function mapPagedAsyncIterable<TGeneratedElement, TMappedElement, TOptions>(
  options: TOptions,
  operation: (options: TOptions & { maxresults?: number }) => GeneratedPagedAsyncIterableIterator<TGeneratedElement>,
  mapper: (x: TGeneratedElement) => TMappedElement
): CorePagedAsyncIterableIterator<TMappedElement> {
  let iter: GeneratedPagedAsyncIterableIterator<TGeneratedElement> | undefined = undefined;
  return {
    async next() {
      iter ??= operation({ ...options, maxresults: undefined });
      const result = await iter.next();

      return {
        ...result,
        value: result.value && mapper(result.value),
      };
    },
    [Symbol.asyncIterator]() {
      return this;
    },
    async *byPage(settings) {
      // note that we defer actually calling the operation to get the iterator until we have the maxPageSize value available
      const iteratorByPage = operation({ ...options, maxresults: settings?.maxPageSize }).byPage(settings);
      for await (const page of iteratorByPage) {
        yield page.map(mapper);
      }
    },
  };
}
```

which could then be used in the convenience client like

```ts
  public listPropertiesOfSecrets(
    options: ListPropertiesOfSecretsOptions = {},
  ): PagedAsyncIterableIterator<SecretProperties> {
    return mapPagedAsyncIterable(
      options,
      updatedOptions => this.client.getSecrets(updatedOptions),
      (item) => getSecretFromSecretBundle(item).properties,
    );
  }
```

If it's possible to wrap like this, I would favor that over a codegen approach, at least for now. Key Vault is likely a special case here and I think it would be good to understand how other libraries use and handle `maxPageSize` before we commit to a more general solution in the generator.

**Comment by xirzec** [comment]
Inconsistencies between services about if page size is client driven or service driven is a huge part of the complexity here, in addition to the parameter names not being consistent. Perhaps TypeSpec can decorate us out of this corner?

**Comment by maorleger** [comment]
Yeah, good point. This is where I'm slowly landing on with keyvault keys: https://github.com/Azure/azure-sdk-for-js/pull/31845/files#diff-381682f38b35959da24d3ed01ec6ceae9b9a4a61184f2d1db49c7ea99bb79e4dR189-R232 if you're curious. 

My plan is to update the 4 libraries and merge them, then go back and move any common helpers (like the above) into keyvault-common. Then, we can revisit the compatibility gap more holistically - it's possible that this is just keyvault quirks...
